server/policy.go

```diff
@@ -92,4 +100,119 @@ func (s *Server) registerPolicyRoutes(g *echo.Group) {
+		if err := api.ValidatePolicy(pType, ""); err != nil {
+			return echo.NewHTTPError(http.StatusBadRequest, fmt.Sprintf("Invalid policy type: %q", pType)).SetInternal(err)
+		}
+
+		policyFind := &api.PolicyFind{
+			Type: &pType,
+		}
+
+		ctx := c.Request().Context()
+		policyList, err := s.store.ListPolicy(ctx, policyFind)
+		if err != nil {
+			return echo.NewHTTPError(http.StatusInternalServerError, fmt.Sprintf("Failed to list policy for type %q", pType)).SetInternal(err)
+		}
+
+		c.Response().Header().Set(echo.HeaderContentType, echo.MIMEApplicationJSONCharsetUTF8)
+		if err := jsonapi.MarshalPayload(c.Response().Writer, policyList); err != nil {
+			return echo.NewHTTPError(http.StatusInternalServerError, fmt.Sprintf("Failed to marshal list policy response: %v", pType)).SetInternal(err)
+		}
+		return nil
+	})
+
+	g.PATCH("/policy/:policyID", func(c echo.Context) error {
+		id, err := strconv.Atoi(c.Param("policyID"))
+		if err != nil {
+			return echo.NewHTTPError(http.StatusBadRequest, fmt.Sprintf("ID is not a number: %s", c.Param("policyID"))).SetInternal(err)
+		}
+
+		policyPatch := &api.PolicyPatch{}
+		if err := jsonapi.UnmarshalPayload(c.Request().Body, policyPatch); err != nil {
+			return echo.NewHTTPError(http.StatusBadRequest, "Malformatted set policy request").SetInternal(err)
+		}
+		policyPatch.ID = id
+		policyPatch.Type = api.PolicyType(c.QueryParam("type"))
+		policyPatch.UpdaterID = c.Get(getPrincipalIDContextKey()).(int)
+
+		payload := ""
+		if policyPatch.Payload != nil {
+			payload = *policyPatch.Payload
+		}
+		if err := s.hasAccessToUpsertPolicy(policyPatch.Type, payload); err != nil {
+			return echo.NewHTTPError(http.StatusForbidden, err.Error()).SetInternal(err)
+		}
+
+		ctx := c.Request().Context()
+		policy, err := s.store.PatchPolicy(ctx, policyPatch)
+		if err != nil {
+			if common.ErrorCode(err) == common.Invalid {
+				return echo.NewHTTPError(http.StatusBadRequest, err.Error()).SetInternal(err)
+			}
+			return echo.NewHTTPError(http.StatusInternalServerError, fmt.Sprintf("Failed to patch policy for type %q", policyPatch.Type)).SetInternal(err)
+		}
+
+		c.Response().Header().Set(echo.HeaderContentType, echo.MIMEApplicationJSONCharsetUTF8)
+		if err := jsonapi.MarshalPayload(c.Response().Writer, policy); err != nil {
+			return echo.NewHTTPError(http.StatusInternalServerError, "Failed to marshal create set policy response").SetInternal(err)
+		}
+		return nil
+	})
+
+	g.DELETE("/policy/:policyID", func(c echo.Context) error {
+		id, err := strconv.Atoi(c.Param("policyID"))
+		if err != nil {
+			return echo.NewHTTPError(http.StatusBadRequest, fmt.Sprintf("ID is not a number: %s", c.Param("policyID"))).SetInternal(err)
+		}
+
+		policyDelete := &api.PolicyDelete{
+			ID:        id,
+			DeleterID: c.Get(getPrincipalIDContextKey()).(int),
+			Type:      api.PolicyType(c.QueryParam("type")),
+		}
+		if err := s.hasAccessToUpsertPolicy(policyDelete.Type, ""); err != nil {
```

Comment:
this can be moved before L171, similar in other places.

Normally do the validation before other code.